### PR TITLE
propogate cite_reference node classes to pending_xref node

### PR DESCRIPTION
**Pull request by chrisjsewell**
Subject: propogate `cite_reference` node classes to `pending_xref` node

### Feature

small addition to `sphinx.transforms.CitationReferences` to
propogate `cite_reference` node classes to `pending_xref` node

### Purpose

This is useful for extensions, 
that wish to identify and act only on `cite_reference` nodes that they create

### CODE REVIEW

File: sphinx/transforms/__init__.py [hunk 1]
```diff
@@ -220,6 +220,8 @@ def apply(self):
             refnode.source = citnode.source or citnode.parent.source
             refnode.line = citnode.line or citnode.parent.line
             refnode += nodes.Text('[' + cittext + ']')
+            for class_name in citnode.attributes.get('classes', []):
+                refnode.set_class(class_name)
```
Comment (tk0miya) [inline sphinx/transforms/__init__.py]: `Element.set_class()` is marked as deprecated. So please modify `refnode['classes']` directly instead.
https://repo.or.cz/docutils.git/blob/HEAD:/docutils/docutils/nodes.py#l1021

**Review by tk0miya — CHANGES_REQUESTED** [review]
LGTM with nits.

**Comment by tk0miya** [comment]
Could you rebase this to 2.0 branch? I'd like to merge into the branch.

**Comment by chrisjsewell** [comment]
@tk0miya the only issue I found while creating PR mcmtroffaes/sphinxcontrib-bibtex#161
is that, if a citation is not found for a citation reference, then in
[this line](https://github.com/sphinx-doc/sphinx/blob/c5f533d73ddc84d0a07ea2f91f9443dd11ec1643/sphinx/transforms/post_transforms/__init__.py#L72)   `docutils.nodes.replace_self` will raise an `AssertionError`, because  `sphinx.addnodes.pending_xref` now has class attributes that cannot be transferred to a `docutils.nodes.Text` node.

Would it be best then to add below [this line](https://github.com/sphinx-doc/sphinx/blob/c5f533d73ddc84d0a07ea2f91f9443dd11ec1643/sphinx/transforms/post_transforms/__init__.py#L42) or [this line](https://github.com/sphinx-doc/sphinx/blob/c5f533d73ddc84d0a07ea2f91f9443dd11ec1643/sphinx/transforms/post_transforms/__init__.py#L71):

```python
contnode['classes'] = []
```

**Comment by tk0miya** [comment]
Thank you for letting me know that. I think the `AttributeError` must be a bug of Sphinx, not bibtex extension's.
Sphinx expects that a child node of pending_xref is kind of TextElement. But our `CitationReferences` transform breaks it. I'll work for it in another PR.
Thanks,

**Review by tk0miya — APPROVED** [review]

**Comment by tk0miya** [comment]
I just made a fix #6151 for `CitationReferences`. Could you check it please?

**Pull request merged**